### PR TITLE
#15 implement destination card component

### PR DESCRIPTION
**Pull request by shizuko-akamoto**
#### `Image` Component
* Added to generalize what to display when no image src is given. The component takes in as props img src, and if that is set to null, will show an empty gray background with `?` icon in the middle.

#### `RatingBar` Component
* RatingBar displays a set of stars, and depending on the `Rating` value (a number from 0~5) as props, will display solid filled stars for `Rating` number of stars, and open stars for the rest.

#### `DestinationCard` Component
* This is a component for one destination in Trecipe. It needs to take in as props a `DestinationModel`, a model representing one destination. 
* Aside from DestinationModel, it also takes in a boolean `isInEdit` to manage the style of the card.
If `isInEdit` == false, a checkbox for checking off the destination is displayed, otherwise, the two edit buttons (Reorder and Delete) are given.
* DC elements are wrapped inside `Draggable`, this is from react-beautiful-dnd, and basically, it makes whatever inside this component draggable. 

#### `Trecipe` Page
* For this PR, Trecipe page only contains a list of destinations (no header, no map, etc). 
* `DragDropContext` and `Droppable` are also components from react-beautiful-dnd, the first is a wrapper for all dnd related components, and the latter, marking area in the web `Draggable` elements can be dropped to.
  * The `reorder` method is called whenever the user drops DC to a new location. This method reorders destinations stored in Trecipe state, and thus list is reordered on next render.
* You will see some initial setup of Trecipe page, like code for "loading" (though just getting 10 copies of mock destination) destinations to page, as well as logic to limit how many destinations in the list are shown. The default initial is set to 5 destination cards right now, and each time "Show More" button is clicked, 5 more appears. Some more rules regarding the "Show More" button:
  * If all DCs are already visible, button is hidden (handled by `canExpand()`)
  * If `isInEdit` == true, we show all the DCs and hide the "Show More" button (handled by `canExpand()` and `getDestinationsList()`

### Demo
DC cards not in edit with "Show More" button:
Trecipe-Google-Chrome-2020-06-16-01-36-15

DC cards in edit with reordering:
Trecipe-Google-Chrome-2020-06-16-01-35-25

### CODE REVIEW

**Comment by tianyan-zhu** [comment]
Design choice: 
Do we want the corners to be a little bit more round? Or do we want to stick with the pointy ones? 

Personally I think we should be consistent with the buttons.

File: src/pages/Trecipe/DestinationCard/DestinationCard.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+/**
+ * DestinationModal
+ * id: destination unique id
+ * name: destination name
+ * category: destination category
+ * rating: destination rating
+ * imgSrc: destination image
+ */
```
Comment (tianyan-zhu) [inline src/pages/Trecipe/DestinationCard/DestinationCard.tsx:25]: minor: address and description missing in your comments.

**Comment by shizuko-akamoto** [comment]
> Design choice:
> Do we want the corners to be a little bit more round? Or do we want to stick with the pointy ones?
> 
> Personally I think we should be consistent with the buttons.

Yea, we can go with 8px border-radius for our boxed conponents 👍

File: src/pages/Trecipe/DestinationCard/DestinationCard.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+export interface DestinationModel {
+  id: number;
+  name: string;
+  category: DestinationCategory;
+  address: string;
+  rating: Rating;
+  description: string;
+  imgSrc: string;
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply src/pages/Trecipe/DestinationCard/DestinationCard.tsx:35]: Good point 👍

File: src/pages/Trecipe/DestinationCard/DestinationCard.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        draggableId={String(this.props.destModel.id)}
+        index={this.props.index}
+        isDragDisabled={!this.props.isInEdit}>
+        {(provided) => (
+          <div ref={provided.innerRef} {...provided.draggableProps}>
+            <div className="dest-card-wrapper">
+              <div className="dest-img">
+                <Image
+                  src={this.props.destModel.imgSrc}
+                  imgStyle={{ borderRadius: "8px 0 0 8px" }}
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply src/pages/Trecipe/DestinationCard/DestinationCard.tsx:83]: This one is actually a Image component prop haha. It's for whatever additional css we wanna apply to the img element inside Image component.

File: src/pages/Trecipe/DestinationCard/DestinationCard.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                <p className="dest-description">
+                  {this.props.destModel.description}
+                </p>
+              </div>
+              {!this.props.isInEdit && (
+                <div className="completed-checkbox">
+                  <input
+                    type="checkbox"
+                    id={this.props.destModel.id + "-completed"}
+                    onChange={this.handleCompletedCheck.bind(this)}
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply src/pages/Trecipe/DestinationCard/DestinationCard.tsx:102]: Right. Good catch, thanks!

**Comment by shizuko-akamoto** [comment]
> Awesome, the drag and drop works really well
> 
> Some more design choice: Do we wanna change the categories into icon?
> 
> And I was experimenting with some randomized linear gradient background instead of displaying a question mark icon when a image is not provided, not sure if this looks better 🤷‍♂️
> 
> image

I added the icons, just beside the category text for now 😄  (until we implement tooltips sometime in future...)

**Comment by tianyan-zhu** [comment]
The Coconut Tree is on the wrong line :P 

image

**Comment by shizuko-akamoto** [comment]
> The Coconut Tree is on the wrong line :P
> 
> image

I think those are four different design suggestions haha

**Comment by tianyan-zhu** [comment]
Ahhh I see, my bad :P

**Comment by shizuko-akamoto** [comment]
> > The Coconut Tree is on the wrong line :P
> > image
> 
> I think those are four different design suggestions haha

(which i did not exactly implement. sorry, I don't really wanna implement tooltips right now LOL)

File: src/pages/Trecipe/DestinationCard/DestinationCard.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (knox153) [reply src/pages/Trecipe/DestinationCard/DestinationCard.tsx:105]: Ohh I see, got it 👍

**Comment by knox153** [comment]
> 
> 
> > > The Coconut Tree is on the wrong line :P
> > > image
> > 
> > 
> > I think those are four different design suggestions haha
> 
> (which i did not exactly implement. sorry, I don't really wanna implement tooltips right now LOL)

Haha no worry, we can always defer design changes. Sorry for suggesting these changes so late, these ideas usually come up when I start playing with it

**Comment by shizuko-akamoto** [comment]
> > > > The Coconut Tree is on the wrong line :P
> > > > image
> > > 
> > > 
> > > I think those are four different design suggestions haha
> > 
> > 
> > (which i did not exactly implement. sorry, I don't really wanna implement tooltips right now LOL)
> 
> Haha no worry, we can always defer design changes. Sorry for suggesting these changes so late, these ideas usually come up when I start playing with it



> > > > The Coconut Tree is on the wrong line :P
> > > > image
> > > 
> > > 
> > > I think those are four different design suggestions haha
> > 
> > 
> > (which i did not exactly implement. sorry, I don't really wanna implement tooltips right now LOL)
> 
> Haha no worry, we can always defer design changes. Sorry for suggesting these changes so late, these ideas usually come up when I start playing with it

We should definitely be keeping track of those hehe. Created a folder in our google drive for all those ideas
https://drive.google.com/drive/folders/1EVdh-ycXN6rijsPRLiYrTivdtWMehZk0

**Pull request merged**